apollo-federation/src/sources/connect/json_selection/immutable.rs

```diff
@@ -0,0 +1,56 @@
+use std::clone::Clone;
+use std::rc::Rc;
+
+#[derive(Debug, Clone)]
+pub struct InputPath<T: Clone> {
+    path: Path<T>,
+}
+
+type Path<T> = Option<Rc<AppendPath<T>>>;
+
+#[derive(Debug, Clone)]
+struct AppendPath<T: Clone> {
+    prefix: Path<T>,
+    last: T,
+}
+
+impl<T: Clone> InputPath<T> {
+    pub fn empty() -> InputPath<T> {
+        InputPath { path: None }
+    }
+
+    pub fn append(&self, last: T) -> Self {
+        Self {
+            path: Some(Rc::new(AppendPath {
+                prefix: self.path.clone(),
+                last,
+            })),
+        }
+    }
+
+    pub fn to_vec(&self) -> Vec<T> {
+        // This method needs to be iterative rather than recursive, to be
+        // consistent with the paranoia of the drop method.
+        let mut vec = Vec::new();
+        let mut path = self.path.as_deref();
+        while let Some(p) = path {
+            vec.push(p.last.clone());
+            path = p.prefix.as_deref();
+        }
+        vec.reverse();
+        vec
+    }
+}
+
+impl<T: Clone> Drop for InputPath<T> {
+    fn drop(&mut self) {
+        let mut path = self.path.take();
+        while let Some(rc) = path {
+            if let Ok(mut p) = Rc::try_unwrap(rc) {
+                path = p.prefix.take();
+            } else {
+                break;
```

Comment:
I'm a bit worried that this will memory leak if something holds a reference to an `InputPath` somewhere in the chain will cause the rest of the paths to never get dropped.

Reply:
Well, that's the idea with a persistent stack. It doesn't get dropped while someone still has an active reference to it. Unless you can be more specific about this threat model, it sounds like a normal reference-counting risk, and the alternative is a lot of unnecessary cloning.

Reply:
I _did_ think I saw a persistent memory leak when I was performance testing this branch a couple of weeks ago, but that was before the manual drop impl.

Maybe let's run a quick load test just to make sure it's not a constant memory leak (which would eventually OOM a router)? I had an idea about how we could rework the path a bit and avoid both Rcs and cloning, but it's messy so... let's not go down that road unless we actually have a problem 😁.

@lennyburdette is there anything in the current performance tests that watches for memory usage, which we can lean on with the new connectors tests?